### PR TITLE
Implement Beacon Node gRPC Server Streams

### PR DESCRIPTION
**Pull request by rauljordan**
Hi all,

Currently, the beacon node's RPC server methods are unimplemented and simply mocked in the client tests. This PR will begin by streaming the latest beacon block and latest crystallized state and ensuring the sharding client on the other end is able to receive these values properly.

In production, the beacon node should stream the **canonical** latest block and state to clients, but for now we will be streaming any received, incoming p2p beacon block (at the moment created by the simulator service) to clients.

### Running Instructions

To run, first start from a fresh beacon node data directory. Then:
- Run a geth node as a miner exposing WS, RPC endpoints
- Start a beacon node with the simulator flag with a fresh beacon node datadir
- Start a client connected to this beacon node

You'll be able to experience blocks being accepted by the node, sent to the client via RPC, and state transitions determining proposer responsibilities.

### CODE REVIEW

File: beacon-chain/rpc/service.go [hunk 1]
```diff
@@ -117,12 +131,35 @@ func (s *Service) SignBlock(ctx context.Context, req *pb.SignRequest) (*pb.SignR
...
 // LatestCrystallizedState streams the latest beacon crystallized state.
 func (s *Service) LatestCrystallizedState(req *empty.Empty, stream pb.BeaconService_LatestCrystallizedStateServer) error {
-	// TODO: implement.
-	return errors.New("unimplemented")
+	// Right now, this streams every newly created crystallized state but should only
+	// stream canonical states.
+	for {
+		select {
+		case state := <-s.announcer.CanonicalCrystallizedStateAnnouncement():
+			log.Info("Sending state to RPC clients")
```
Comment (terenc3t) [inline beacon-chain/rpc/service.go]: Can we say `Sending crystallized state to RPC clients` instead? state could be active or crystallized and it could be confusing

File: beacon-chain/types/state.go [hunk 2]
```diff
@@ -37,6 +37,11 @@ func NewGenesisStates() (*ActiveState, *CrystallizedState) {
 			RecentBlockHashes:   [][]byte{},
 		},
 	}
+	var validators []*pb.ValidatorRecord
+	for i := 0; i < 100; i++ {
+		validator := &pb.ValidatorRecord{StartDynasty: 0, EndDynasty: 100, Balance: 1000, WithdrawalAddress: []byte{}, PublicKey: 0}
```
Comment (terenc3t) [inline beacon-chain/types/state.go]: Set `EndDynasty` to `params.DefaultEndDynasty` or we'll run into issues with simulate block advancing and dynasty passes 100 mark

Comment (terenc3t) [reply beacon-chain/types/state.go]: Also `balance` to `params.DefaultBalance`

**Comment by rawfalafel** [comment]
https://github.com/prysmaticlabs/prysm/blob/1500563497abbfae37bbd570ff3f0bc149774f09/beacon-chain/node/node.go#L170

I think this should return an error instead

**Comment by terenc3t** [comment]
I thought about having that returning an error. But it's should still a valid use case for users to launch beacon node without access to PoW chain if they just wanna observe and don't participate in validator duty right? I think it should be log.warn with a message like "can't access web3 service, running observer mode" Does that make sense?

**Comment by rawfalafel** [comment]
Ah, I see. I'm not against that, especially if it makes dev easier. If we do go in that direction, we should tweak [CanProcessBlock](https://github.com/prysmaticlabs/prysm/blob/master/beacon-chain/blockchain/core.go#L188) so that it doesn't always fail.

File: beacon-chain/blockchain/core.go [hunk 3]
```diff
@@ -191,11 +192,14 @@ func (b *BeaconChain) CanProcessBlock(fetcher types.POWBlockFetcher, block *type
...
 	parentHash := block.ParentHash()
+	var genesisHash [32]byte
+	copy(genesisHash[:], []byte("genesis"))
+
 	hasParent, err := b.db.Has(parentHash[:])
 	if err != nil {
 		return false, err
 	}
-	if !hasParent {
+	if !hasParent && !bytes.Equal(parentHash[:], genesisHash[:]) {
```
[the remark this answers is not in the document]
Comment (rauljordan) [reply beacon-chain/blockchain/core.go]: Good idea, way easier

File: beacon-chain/blockchain/core.go [hunk 4]
```diff
@@ -195,7 +195,7 @@ func (b *BeaconChain) CanProcessBlock(fetcher types.POWBlockFetcher, block *type
 	if err != nil {
 		return false, err
 	}
-	if !hasParent {
+	if !hasParent && block.SlotNumber() != 1 {
 		return false, errors.New("parent hash points to nil in beaconDB")
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/core.go]: Can you clarify this? I found it confusing that it was an ERROR level logging and what caused it.

File: beacon-chain/blockchain/service.go [hunk 5]
```diff
@@ -44,15 +48,17 @@ func DefaultConfig() *Config {
...
-		processedBlockHashes:             [][32]byte{},
-		processedActiveStateHashes:       [][32]byte{},
-		processedCrystallizedStateHashes: [][32]byte{},
+		ctx:                                    ctx,
+		chain:                                  beaconChain,
+		cancel:                                 cancel,
+		beaconDB:                               beaconDB,
+		web3Service:                            web3Service,
+		latestBeaconBlock:                      make(chan *types.Block, cfg.BeaconBlockBuf),
+		canonicalBlockAnnouncement:             make(chan *types.Block, cfg.AnnouncementBuf),
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/service.go]: Is there a smaller word you can use other than Announcement? 

In go, we should prefer the smallest variable names possible without losing context or information. Maybe `Event` would be a good substitute?

File: beacon-chain/blockchain/service.go [hunk 6]
```diff
@@ -16,26 +16,30 @@ var log = logrus.WithField("prefix", "blockchain")
...
-	processedCrystallizedStateHashes [][32]byte
+	ctx                                    context.Context
+	cancel                                 context.CancelFunc
+	beaconDB                               *database.DB
+	chain                                  *BeaconChain
+	web3Service                            *powchain.Web3Service
+	canonicalBlockAnnouncement             chan *types.Block
+	canonicalCrystallizedStateAnnouncement chan *types.CrystallizedState
+	latestBeaconBlock                      chan *types.Block
+	processedBlockHashes                   [][32]byte
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/service.go]: What are these used for? Are we keeping a copy of all of the blocks we've seen in memory? 

Some comments on these fields would be helpful for reader understanding.

File: beacon-chain/blockchain/service.go [hunk 7]
```diff
@@ -200,27 +221,43 @@ func (c *ChainService) run(done <-chan struct{}) {
 			validatorsByHeight, err := c.chain.validatorsByHeightShard()
 			if err != nil {
 				log.Errorf("Unable to get validators by height and by shard: %v", err)
+				continue
 			}
-			log.Debugf("Received the following validators by height: %v", validatorsByHeight)
+			log.Debugf("Received the following number of validators by height: %v", len(validatorsByHeight))
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/service.go]: This is odd phrasing. Can you change this to `Received %d validators by height`?

File: beacon-chain/blockchain/service_test.go [hunk 8]
```diff
@@ -87,7 +87,7 @@ func TestStartStop(t *testing.T) {
 	if len(chainService.CurrentActiveState().RecentBlockHashes()) != 0 {
 		t.Errorf("incorrect recent block hashes")
 	}
-	if len(chainService.CurrentCrystallizedState().Validators()) != 0 {
+	if len(chainService.CurrentCrystallizedState().Validators()) != 100 {
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/service_test.go]: What is this magic number 100? Can you add a comment?

File: beacon-chain/types/state.go [hunk 9]
```diff
@@ -37,6 +38,11 @@ func NewGenesisStates() (*ActiveState, *CrystallizedState) {
 			RecentBlockHashes:   [][]byte{},
 		},
 	}
+	var validators []*pb.ValidatorRecord
+	for i := 0; i < 100; i++ {
```
Comment (prestonvanloon) [inline beacon-chain/types/state.go]: Why 100? Where does this number come from?

File: beacon-chain/blockchain/core.go (hunk 4 above)
Comment (rauljordan) [reply beacon-chain/blockchain/core.go]: These should be debug level, basically saying that the new incoming block either had (1) wrong state hashes, (2) has no parent in the db, or (3) has no valid mainchain reference

File: beacon-chain/types/block.go [hunk 10]
```diff
@@ -32,14 +31,10 @@ func NewBlock(data *pb.BeaconBlock) (*Block, error) {
 				ActiveStateHash:       []byte{0},
 				CrystallizedStateHash: []byte{0},
 			},
-		}, nil
+		}
 	}
 
-	if len(data.ParentHash) != 32 {
```
Comment (rawfalafel) [inline beacon-chain/types/block.go]: Is there a reason why this was removed? I asked @s1na to add this check because protobufs can't validate that a byte array is a fixed length, and in practice a malicious peer can send a byte array of any length.

Comment (rauljordan) [reply beacon-chain/types/block.go]: This occurs because protobufs cannot store [32]byte but only []byte, so it is not possible for this condition to be met

File: beacon-chain/rpc/service.go [hunk 11]
```diff
@@ -35,14 +38,15 @@ type Config struct {
 
 // NewRPCService creates a new instance of a struct implementing the BeaconServiceServer
 // interface.
-func NewRPCService(ctx context.Context, cfg *Config) *Service {
+func NewRPCService(ctx context.Context, cfg *Config, announcer types.CanonicalEventAnnouncer) *Service {
```
Comment (prestonvanloon) [inline beacon-chain/rpc/service.go]: Can the announcer be part of the config?

File: beacon-chain/simulator/service.go [hunk 12]
```diff
@@ -43,26 +47,28 @@ type Config struct {
...
-		Delay:                       time.Second * 8,
+		Delay:                       time.Second * 5,
 		BlockRequestBuf:             100,
 		CrystallizedStateRequestBuf: 100,
 	}
 }
 
 // NewSimulator creates a simulator instance for a syncer to consume fake, generated blocks.
-func NewSimulator(ctx context.Context, cfg *Config, beaconp2p types.P2P, web3Service types.POWChainService, chainService types.StateFetcher) *Simulator {
+func NewSimulator(ctx context.Context, cfg *Config, beaconDB ethdb.Database, beaconp2p types.P2P, web3Service types.POWChainService, chainService types.StateFetcher) *Simulator {
```
Comment (prestonvanloon) [inline beacon-chain/simulator/service.go]: 6 args is way too many.

Is there a reason to have so many? If this is out of scope, we can solve in another PR.

File: beacon-chain/rpc/service.go (hunk 11 above)
Comment (rauljordan) [reply beacon-chain/rpc/service.go]: I run into weird conditions when testing because I send to a channel in the announcer interface implementation if I put it as part of config, as it seems it creates a copy that is different in the RPCService struct if I do that

Comment (rauljordan) [reply beacon-chain/rpc/service.go]: We can explore this in a separate PR if this service grows beyond 3 arguments
